vm-migration/src/lib.rs

```diff
@@ -139,6 +108,10 @@ impl SnapshotData {
 /// the actual device snapshot data.
 #[derive(Clone, Default, Deserialize, Serialize)]
 pub struct Snapshot {
+    /// Version of this snapshot: can be used for transitions in the future if
+    /// necessary
+    pub version: u64,
+
```

Comment:
I think I covered it in the previous  comment - but let me try again.

If say - we need to handle an upgrade that needs the version in the virtio-net code - it would need to know that the version of the virtio-net state has changed from version `x` to version `x+1` and adapt appropriately. 

I think that having version on `Snapshot` is useful in case we want to change the layout of the snapshot struct itself..

Reply:
> I think I covered it in the previous comment - but let me try again.
> 
> If say - we need to handle an upgrade that needs the version in the virtio-net code - it would need to know that the version of the virtio-net state has changed from version `x` to version `x+1` and adapt appropriately.
> 

My main point is, assuming the version of `SnapshotData` and `Snapshot` will always stay consistent (e.g. the same), it won't need to be included in both places, say we can always access the version information from the `Snapshot` struct via its `SnapshotData` field. 

> I think that having version on `Snapshot` is useful in case we want to change the layout of the snapshot struct itself..

For such case, I agree having separate version on `Snapshot` itself could be helpful. I'd say such kind of changes would be quite dramatic (indicating major refactoring on the way how we do snapshot/restore). I am not sure if we will be able to keep the live-upgrade in that case even with the version information.

Reply:
I envisioned the version being local to the thing that changed - that's why I gave the particular example of virtio-net.

The version for `SnapshotData` would change if the related `*State` struct changes e.g. if the `NetState` struct changed in a way that needs transition when then the code turning the `SnapshotData` into the `NetState` would need to handle that.

Perhaps the right approach is to add a `version` field to every `State` struct - that would make it more obvious? WDYT?

Reply:
Oh, I see. If that's the purpose of the version information you want, adding it to every `*State` struct is more obvious and better. 

But I don't see how such version information would be helpful for transitions when there are changes to the `*State` struct. Let's say the `NetState` is changed with a new field, and we give it a new version. How would the version bump within the `NetState` struct help with not breaking the live upgrade for the transition time?

Reply:
I've come to the same realisation that having it in the `*State` struct is too late - by that time it's already been deserialized from the `SnapshotData` - so we need the version in the `SnapshotData` so that when deserializing the restore logic can make the correct choice (e.g. deserializing to a different version of the struct or treating it as a dictionary and manually accessing the values for a transition.)

I'm going to take the version stuff out from this PR as I think it needs to be considered carefully and since we ant to expedite this change that seems the best plan.

Reply:
> I don't see how such version information would be helpful for transitions when there are changes to the `*State` struct.

Agree.
If the `*State` struct change, the Restore will report an error such like this:
Error running command: Server responded with an error: InternalServerError: ApiError(VmRestore(Restore(Restore(Could not create devices RestoreGetState(Restore(Error deserialising: _net1-section A deserialization error occured: Io(Error { kind: UnexpectedEof, message: "failed to fill whole buffer" })))))))

Reply:
We have to handle it manually e.g. create if version == 1 deserialize to a e.g. `NetStateV1` struct and copy the values out where appropriate - or if necessary use fields with a serde default - it can be handled on a case by ase basis depending on what is needed for the transition.